File: STM32/STM32-F103ZET6/User/e-Paper/EPD_1in54b_V2.c
/*****************************************************************************
* | File      	:   EPD_1in54b_V2.h
* | Author      :   Waveshare team
* | Function    :   1.54inch e-paper b V2
* | Info        :
*----------------
* |	This version:   V1.0
* | Date        :   2020-04-02
* | Info        :
* -----------------------------------------------------------------------------
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
******************************************************************************/
#include "EPD_1in54b_V2.h"
#include "Debug.h"

/******************************************************************************
function :	Software reset
parameter:
******************************************************************************/
static void EPD_1IN54B_V2_Reset(void)
{
    DEV_Digital_Write(EPD_RST_PIN, 1);
    DEV_Delay_ms(100);
    DEV_Digital_Write(EPD_RST_PIN, 0);
    DEV_Delay_ms(2);
    DEV_Digital_Write(EPD_RST_PIN, 1);
    DEV_Delay_ms(100);
}

/******************************************************************************
function :	send command
parameter:
     Reg : Command register
******************************************************************************/
static void EPD_1IN54B_V2_SendCommand(UBYTE Reg)
{
    DEV_Digital_Write(EPD_DC_PIN, 0);
    DEV_Digital_Write(EPD_CS_PIN, 0);
    DEV_SPI_WriteByte(Reg);
    DEV_Digital_Write(EPD_CS_PIN, 1);
}

/******************************************************************************
function :	send data
parameter:
    Data : Write data
******************************************************************************/
static void EPD_1IN54B_V2_SendData(UBYTE Data)
{
    DEV_Digital_Write(EPD_DC_PIN, 1);
    DEV_Digital_Write(EPD_CS_PIN, 0);
    DEV_SPI_WriteByte(Data);
    DEV_Digital_Write(EPD_CS_PIN, 1);
}

/******************************************************************************
function :	Wait until the busy_pin goes LOW
parameter:
******************************************************************************/
static void EPD_1IN54B_V2_ReadBusy(void)
{
    Debug("e-Paper busy\r\n");
    while(1) {
        if(DEV_Digital_Read(EPD_BUSY_PIN) == 0)
            break;
    }
    DEV_Delay_ms(200);
    Debug("e-Paper busy release\r\n");
}


/******************************************************************************
function :	Initialize the e-Paper register
parameter:
******************************************************************************/
void EPD_1IN54B_V2_Init(void)
{
    EPD_1IN54B_V2_Reset();

    EPD_1IN54B_V2_ReadBusy();   
    EPD_1IN54B_V2_SendCommand(0x12);  //SWRESET
    EPD_1IN54B_V2_ReadBusy();   

    EPD_1IN54B_V2_SendCommand(0x01); //Driver output control      
    EPD_1IN54B_V2_SendData(0xC7);
    EPD_1IN54B_V2_SendData(0x00);
    EPD_1IN54B_V2_SendData(0x01);

    EPD_1IN54B_V2_SendCommand(0x11); //data entry mode       
    EPD_1IN54B_V2_SendData(0x01);

    EPD_1IN54B_V2_SendCommand(0x44); //set Ram-X address start/end position   
    EPD_1IN54B_V2_SendData(0x00);
    EPD_1IN54B_V2_SendData(0x18);    //0x18-->(24+1)*8=200

    EPD_1IN54B_V2_SendCommand(0x45); //set Ram-Y address start/end position          
    EPD_1IN54B_V2_SendData(0xC7);    //0xC7-->(199+1)=200
    EPD_1IN54B_V2_SendData(0x00);
    EPD_1IN54B_V2_SendData(0x00);
    EPD_1IN54B_V2_SendData(0x00); 

    EPD_1IN54B_V2_SendCommand(0x3C); //BorderWavefrom
    EPD_1IN54B_V2_SendData(0x05);

    EPD_1IN54B_V2_SendCommand(0x18); //Read built-in temperature sensor
    EPD_1IN54B_V2_SendData(0x80);

    EPD_1IN54B_V2_SendCommand(0x4E);   // set RAM x address count to 0;
    EPD_1IN54B_V2_SendData(0x00);
    EPD_1IN54B_V2_SendCommand(0x4F);   // set RAM y address count to 0X199;    
    EPD_1IN54B_V2_SendData(0xC7);
    EPD_1IN54B_V2_SendData(0x00);
    EPD_1IN54B_V2_ReadBusy();
}

/******************************************************************************
function :	Clear screen
parameter:
******************************************************************************/
void EPD_1IN54B_V2_Clear(void)
{
    
    unsigned int i;	
    EPD_1IN54B_V2_SendCommand(0x24);   //write RAM for black(0)/white (1)
    for(i=0;i<5000;i++)
    {               
        EPD_1IN54B_V2_SendData(0xff);
    }
    EPD_1IN54B_V2_SendCommand(0x26);   //write RAM for black(0)/white (1)
    for(i=0;i<5000;i++)
    {               
        EPD_1IN54B_V2_SendData(0x00);
    }
    EPD_1IN54B_V2_SendCommand(0x22); //Display Update Control
    EPD_1IN54B_V2_SendData(0xF7);   
    EPD_1IN54B_V2_SendCommand(0x20);  //Activate Display Update Sequence
    EPD_1IN54B_V2_ReadBusy();   
}

/******************************************************************************
function :	Sends the image buffer in RAM to e-Paper and displays
parameter:
******************************************************************************/
void EPD_1IN54B_V2_Display(const UBYTE *blackimage, const UBYTE *redimage)
{
    UWORD Width, Height;
    Width = (EPD_1IN54B_V2_WIDTH % 8 == 0)? (EPD_1IN54B_V2_WIDTH / 8 ): (EPD_1IN54B_V2_WIDTH / 8 + 1);
    Height = EPD_1IN54B_V2_HEIGHT;

     unsigned int i;	
    EPD_1IN54B_V2_SendCommand(0x24);   //write RAM for black(0)/white (1)
    for(i=0;i<Width * Height;i++)
    {               
        EPD_1IN54B_V2_SendData(blackimage[i]);
    }
    EPD_1IN54B_V2_SendCommand(0x26);   //write RAM for black(0)/white (1)
    for(i=0;i<Width * Height;i++)
    {               
        EPD_1IN54B_V2_SendData(~redimage[i]);
    }
    EPD_1IN54B_V2_SendCommand(0x22); //Display Update Control
    EPD_1IN54B_V2_SendData(0xF7);   
    EPD_1IN54B_V2_SendCommand(0x20);  //Activate Display Update Sequence
    EPD_1IN54B_V2_ReadBusy();   
}

/******************************************************************************
function :	Enter sleep mode
parameter:
******************************************************************************/
void EPD_1IN54B_V2_Sleep(void)
{
    EPD_1IN54B_V2_SendCommand(0x10); //enter deep sleep
    EPD_1IN54B_V2_SendData(0x01); 
    DEV_Delay_ms(100);
}

File: RaspberryPi_JetsonNano/python/lib/waveshare_epd/epd1in54b_V2.py
# *****************************************************************************
# * | File        :	  epd1in54b.py
# * | Author      :   Waveshare team
# * | Function    :   Electronic paper driver
# * | Info        :
# *----------------
# * | This version:   V4.0
# * | Date        :   2019-06-20
# # | Info        :   python demo
# -----------------------------------------------------------------------------
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documnetation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to  whom the Software is
# furished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS OR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

import logging
from . import epdconfig

# Display resolution
EPD_WIDTH       = 200
EPD_HEIGHT      = 200

logger = logging.getLogger(__name__)

class EPD:
    def __init__(self):
        self.reset_pin = epdconfig.RST_PIN
        self.dc_pin = epdconfig.DC_PIN
        self.busy_pin = epdconfig.BUSY_PIN
        self.cs_pin = epdconfig.CS_PIN
        self.width = EPD_WIDTH
        self.height = EPD_HEIGHT


    # Hardware reset
    def reset(self):
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(200) 
        epdconfig.digital_write(self.reset_pin, 0) # module reset
        epdconfig.delay_ms(5)
        epdconfig.digital_write(self.reset_pin, 1)
        epdconfig.delay_ms(200)   

    def send_command(self, command):
        epdconfig.digital_write(self.dc_pin, 0)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([command])
        epdconfig.digital_write(self.cs_pin, 1)

    def send_data(self, data):
        epdconfig.digital_write(self.dc_pin, 1)
        epdconfig.digital_write(self.cs_pin, 0)
        epdconfig.spi_writebyte([data])
        epdconfig.digital_write(self.cs_pin, 1)
        
    def ReadBusy(self):
        logger.debug("e-Paper busy")
        while(epdconfig.digital_read(self.busy_pin) == 1):
            epdconfig.delay_ms(100)    
        logger.debug("e-Paper busy release")
        
    def init(self):
        if (epdconfig.module_init() != 0):
            return -1
        # EPD hardware init start
        self.reset()
        
        self.ReadBusy()   
        self.send_command(0x12)  #SWRESET
        self.ReadBusy()   

        self.send_command(0x01) #Driver output control      
        self.send_data(0xC7)
        self.send_data(0x00)
        self.send_data(0x01)

        self.send_command(0x11) #data entry mode       
        self.send_data(0x01)

        self.send_command(0x44) #set Ram-X address start/end position   
        self.send_data(0x00)
        self.send_data(0x18)    #0x18-->(24+1)*8=200

        self.send_command(0x45) #set Ram-Y address start/end position          
        self.send_data(0xC7)    #0xC7-->(199+1)=200
        self.send_data(0x00)
        self.send_data(0x00)
        self.send_data(0x00) 

        self.send_command(0x3C) #BorderWavefrom
        self.send_data(0x05)

        self.send_command(0x18) #Read built-in temperature sensor
        self.send_data(0x80)

        self.send_command(0x4E)   # set RAM x address count to 0
        self.send_data(0x00)
        self.send_command(0x4F)   # set RAM y address count to 0X199    
        self.send_data(0xC7)
        self.send_data(0x00)
        self.ReadBusy()
        return 0

    def getbuffer(self, image):
        buf = [0xFF] * int(self.width * self.height / 8)
        # Set buffer to value of Python Imaging Library image.
        # Image must be in mode 1.
        image_monocolor = image.convert('1')
        imwidth, imheight = image_monocolor.size
        if imwidth != self.width or imheight != self.height:
            raise ValueError('Image must be same dimensions as display \
                ({0}x{1}).' .format(self.width, self.height))

        pixels = image_monocolor.load()
        for y in range(self.height):
            for x in range(self.width):
                # Set the bits for the column of pixels at the current position.
                if pixels[x, y] == 0:
                    buf[int((x + y * self.width) / 8)] &= ~(0x80 >> (x % 8))
        return buf

    def display(self, blackimage, redimage):
        # send black data
        if (blackimage != None):
            self.send_command(0x24) # DATA_START_TRANSMISSION_1
            for i in range(0, int(self.width * self.height / 8)):
                self.send_data(blackimage[i])
                
        # send red data        
        if (redimage != None):
            self.send_command(0x26) # DATA_START_TRANSMISSION_2
            for i in range(0, int(self.width * self.height / 8)):
                self.send_data(~redimage[i])  

        self.send_command(0x22) # DISPLAY_REFRESH
        self.send_data(0xF7)
        self.send_command(0x20) # DISPLAY_REFRESH
        self.ReadBusy()

    def Clear(self):
        self.send_command(0x24) # DATA_START_TRANSMISSION_1
        for i in range(0, int(self.width * self.height / 8)):
            self.send_data(0xFF)
            
        self.send_command(0x26) # DATA_START_TRANSMISSION_2
        for i in range(0, int(self.width * self.height / 8)):
            self.send_data(0x00)

        self.send_command(0x22) # DISPLAY_REFRESH
        self.send_data(0xF7)
        self.send_command(0x20) # DISPLAY_REFRESH
        self.ReadBusy()


    def sleep(self):
        self.send_command(0x10) #enter deep sleep
        self.send_data(0x01) 

        epdconfig.delay_ms(2000)
        epdconfig.module_exit()

### END OF FILE ###

